### PR TITLE
Data nodes sorted

### PR DESCRIPTION
**Pull request by lePici**
# Overview of changes

1. Plugin type 'lyds_tree' is integrated. Used in metadata and it is
internal, which is distinguished by the unset print callback.
Internal metadata must not be printed.
2. Leaf-list and list data instances are now automatically sorted.
3. Both the explicit and implicit YANG statement 'ordered-by system'
will cause the nodes to be ordered.
4. The first instance of the list or leaf-list of node contain
new metadata named 'lyds_tree'.
5. Data nodes are sorted only if their data types have
callback 'sort' set.
6. Sorting can be turned off so far only by adding the 'ordered-by user'
statement to the YANG model.
7. If the sort fails for some reason, the node is still inserted
as the last instance. A warning message informs about this situation.
8. The time required for sorting should be relatively small thanks to
the Red-black tree implementation.
9. Memory requirements will now increase by 40 bytes per list/leaf-list
data node, plus metadata structure overhead.

# perf

## memory

test: 'create new text' without leaf-list inserts

1000 nodes
devel -     3,177,195
lyds_tree - 3,220,116
difference: 42,921

100000 nodes
devel -     122,591,019
lyds_tree - 126,593,940
difference: 4,002,921

40 bytes per lyd_node + metadata overhead

## branch lyds_tree

### 1000-10

```
ly_perf:                                                                                                                                         
        data set size: 1000                                                                                                                      
        each test executed: 10 times                                                                                                             
                                                                                                                                                 
| create new text .................. | 0.002257 s |                                                                                              
| create new bin ................... | 0.001277 s |                                                                                              
| create path ...................... | 0.004225 s |                                                                                              
| validate ......................... | 0.000314 s |                                                                                              
| parse xml mem validate ........... | 0.003583 s |                                                                                              
| parse xml mem no validate ........ | 0.002826 s |                                                                                              
| parse xml file no validate format  | 0.002939 s |                                                                                              
| parse json mem validate .......... | 0.002679 s |                                                                                              
| parse json mem no validate ....... | 0.001798 s |                                                                                              
| parse json file no validate format | 0.001958 s |                                                                                              
| parse lyb mem validate ........... | 0.002584 s |                                                                                              
| parse lyb mem no validate ........ | 0.001783 s |                                                                                              
| parse lyb file no validate ....... | 0.001830 s |                                                                                              
| print xml ........................ | 0.001235 s |                                                                                              
| print json ....................... | 0.000796 s |                                                                                              
| print lyb ........................ | 0.000626 s |                                                                                              
| dup .............................. | 0.000496 s |                                                                                              
| free ............................. | 0.000513 s |                                                                                              
| xpath find ....................... | 0.000731 s |                                                                                              
| xpath find hash .................. | 0.000005 s |                                                                                              
| compare same ..................... | 0.000134 s |                                                                                              
| diff same ........................ | 0.001968 s |                                                                                              
| diff no same ..................... | 0.002004 s |                                                                                              
| merge same ....................... | 0.000978 s |                                                                                              
| merge no same .................... | 0.001157 s |                                                                                              
| merge no same destruct ........... | 0.000640 s |
```

### 100000-3

```
ly_perf:                                                                                                                                         
        data set size: 100000                                                                                                                    
        each test executed: 3 times                                                                                                              
                                                                                                                                                 
| create new text .................. | 0.249784 s |                                                                                              
| create new bin ................... | 0.158706 s |                                                                                              
| create path ...................... | 0.447686 s |                                                                                              
| validate ......................... | 0.049997 s |                                                                                              
| parse xml mem validate ........... | 0.459106 s |                                                                                              
| parse xml mem no validate ........ | 0.328475 s |                                                                                              
| parse xml file no validate format  | 0.342890 s |                                                                                              
| parse json mem validate .......... | 0.348044 s |                                                                                              
| parse json mem no validate ....... | 0.227186 s |                                                                                              
| parse json file no validate format | 0.244726 s |                                                                                              
| parse lyb mem validate ........... | 0.338318 s |                                                                                              
| parse lyb mem no validate ........ | 0.215666 s |                                                                                              
| parse lyb file no validate ....... | 0.215394 s |                                                                                              
| print xml ........................ | 0.119830 s |                                                                                              
| print json ....................... | 0.090757 s |                                                                                              
| print lyb ........................ | 0.053704 s |                                                                                              
| dup .............................. | 0.097857 s |                                                                                              
| free ............................. | 0.104160 s |                                                                                              
| xpath find ....................... | 0.085618 s |                                                                                              
| xpath find hash .................. | 0.000011 s |                                                                                              
| compare same ..................... | 0.027562 s |                                                                                              
| diff same ........................ | 0.309946 s |                                                                                              
| diff no same ..................... | 0.329439 s |                                                                                              
| merge same ....................... | 0.159534 s |                                                                                              
| merge no same .................... | 0.225557 s |                                                                                              
| merge no same destruct ........... | 0.110938 s |
```

## branch devel

### 1000-10

```
ly_perf:                                                                                                                                         
        data set size: 1000
        each test executed: 10 times

| create new text .................. | 0.002367 s |
| create new bin ................... | 0.001393 s |
| create path ...................... | 0.004338 s |
| validate ......................... | 0.000333 s |
| parse xml mem validate ........... | 0.003538 s |
| parse xml mem no validate ........ | 0.002913 s |
| parse xml file no validate format  | 0.003118 s |
| parse json mem validate .......... | 0.002562 s |
| parse json mem no validate ....... | 0.001946 s |
| parse json file no validate format | 0.002321 s |
| parse lyb mem validate ........... | 0.002511 s |
| parse lyb mem no validate ........ | 0.001914 s |
| parse lyb file no validate ....... | 0.001943 s |
| print xml ........................ | 0.001435 s |
| print json ....................... | 0.001082 s |
| print lyb ........................ | 0.000596 s |
| dup .............................. | 0.000627 s |
| free ............................. | 0.000484 s |
| xpath find ....................... | 0.000865 s |
| xpath find hash .................. | 0.000007 s |
| compare same ..................... | 0.000148 s |
| diff same ........................ | 0.002134 s |
| diff no same ..................... | 0.002258 s |
| merge same ....................... | 0.001115 s |
| merge no same .................... | 0.001104 s |
| merge no same destruct ........... | 0.000554 s |
```

### 100000-3

```
ly_perf:                                                                                                                                         
        data set size: 100000                                                                                                                    
        each test executed: 3 times                                                                                                              
                                                                                                                                                 
| create new text .................. | 0.203631 s |                                                                                              
| create new bin ................... | 0.130454 s |                                                                                              
| create path ...................... | 0.411399 s |                                                                                              
| validate ......................... | 0.049272 s |                                                                                              
| parse xml mem validate ........... | 0.413775 s |                                                                                              
| parse xml mem no validate ........ | 0.340388 s |                                                                                              
| parse xml file no validate format  | 0.349116 s |                                                                                              
| parse json mem validate .......... | 0.318907 s |                                                                                              
| parse json mem no validate ....... | 0.238310 s |                                                                                              
| parse json file no validate format | 0.257953 s |                                                                                              
| parse lyb mem validate ........... | 0.302686 s |                                                                                              
| parse lyb mem no validate ........ | 0.232348 s |                                                                                              
| parse lyb file no validate ....... | 0.230928 s |                                                                                              
| print xml ........................ | 0.122265 s |                                                                                              
| print json ....................... | 0.088346 s |                                                                                              
| print lyb ........................ | 0.054410 s |                                                                                              
| dup .............................. | 0.103008 s |                                                                                              
| free ............................. | 0.077809 s |                                                                                              
| xpath find ....................... | 0.087260 s |                                                                                              
| xpath find hash .................. | 0.000013 s |                                                                                              
| compare same ..................... | 0.030518 s |                                                                                              
| diff same ........................ | 0.316554 s |                                                                                              
| diff no same ..................... | 0.329235 s |                                                                                              
| merge same ....................... | 0.164872 s |                                                                                              
| merge no same .................... | 0.201501 s |                                                                                              
| merge no same destruct ........... | 0.089045 s |
```

### CODE REVIEW

File: src/plugins_types/identityref.c [hunk 1]
```diff
@@ -301,14 +301,25 @@ lyplg_type_store_identityref(const struct ly_ctx *ctx, const struct lysc_type *t
...
         return LY_SUCCESS;
     }
     return LY_ENOT;
 }
 
+LIBYANG_API_DEF int
+lyplg_type_sort_identityref(const struct ly_ctx *UNUSED(ctx), const struct lyd_value *val1,
+        const struct lyd_value *val2)
+{
+    int cmp;
```
Comment (michalvasko) [inline src/plugins_types/identityref.c:317]: Redundant variable.

File: src/plugins_types/ipv4_address.c [hunk 2]
```diff
@@ -224,6 +226,36 @@ lyplg_type_compare_ipv4_address(const struct lyd_value *val1, const struct lyd_v
...
+        return cmp;
+    }
+
+    if (!v1->zone && v2->zone) {
+        return -1;
+    } else if (v1->zone && !v2->zone) {
+        return 1;
+    } else if (v1->zone && v2->zone) {
+        cmp = strcmp(v1->zone, v2->zone);
+        return cmp;
```
Comment (michalvasko) [inline src/plugins_types/ipv4_address.c:253]: Just `return strcmp(v1->zone, v2->zone);`.

File: src/plugins_types/ipv4_address_no_zone.c [hunk 3]
```diff
@@ -141,6 +144,24 @@ lyplg_type_compare_ipv4_address_no_zone(const struct lyd_value *val1, const stru
...
 
+/**
+ * @brief Implementation of ::lyplg_type_sort_clb for the ipv4-address-no-zone ietf-inet-types type.
+ */
+static int
+lyplg_type_sort_ipv4_address_no_zone(const struct ly_ctx *UNUSED(ctx), const struct lyd_value *val1,
+        const struct lyd_value *val2)
+{
+    struct lyd_value_ipv4_address_no_zone *v1, *v2;
+    int cmp;
```
Comment (michalvasko) [inline src/plugins_types/ipv4_address_no_zone.c:155]: Redundant var.

File: src/plugins_types/ipv4_prefix.c [hunk 4]
```diff
@@ -212,6 +215,24 @@ lyplg_type_compare_ipv4_prefix(const struct lyd_value *val1, const struct lyd_va
...
 
+/**
+ * @brief Implementation of ::lyplg_type_sort_clb for the ietf-inet-types ipv4-prefix type.
+ */
+static int
+lyplg_type_sort_ipv4_prefix(const struct ly_ctx *UNUSED(ctx), const struct lyd_value *val1,
+        const struct lyd_value *val2)
+{
+    struct lyd_value_ipv4_prefix *v1, *v2;
+    int cmp;
```
Comment (michalvasko) [inline src/plugins_types/ipv4_prefix.c:226]: Redundant var.

File: src/plugins_types/ipv6_prefix.c [hunk 5]
```diff
@@ -226,6 +229,24 @@ lyplg_type_compare_ipv6_prefix(const struct lyd_value *val1, const struct lyd_va
...
 
+/**
+ * @brief Implementation of ::lyplg_type_sort_clb for the ietf-inet-types ipv6-prefix type.
+ */
+static int
+lyplg_type_sort_ipv6_prefix(const struct ly_ctx *UNUSED(ctx), const struct lyd_value *val1,
+        const struct lyd_value *val2)
+{
+    struct lyd_value_ipv6_prefix *v1, *v2;
+    int cmp;
```
Comment (michalvasko) [inline src/plugins_types/ipv6_prefix.c:240]: Redundant.

File: src/plugins_types/lyds_tree.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        struct lys_glob_unres *UNUSED(unres), struct ly_err_item **UNUSED(err))
+{
+    int ret;
+    struct rb_node *rbt = NULL;
+    struct lyd_node *leader;
+    struct lyd_value_lyds_tree *val = NULL;
+
+    if (format == LY_VALUE_CANON) {
+        /* The canonical value for lyds_tree type is the empty string, so @p value is like NULL. */
+        memset((storage)->fixed_mem, 0, LYD_VALUE_FIXED_MEM_SIZE);
```
Comment (michalvasko) [inline src/plugins_types/lyds_tree.c:41]: Redundant parentheses.

File: src/plugins_types/lyds_tree.c [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+static LY_ERR
+lyplg_type_dupl_lyds(const struct ly_ctx *UNUSED(ctx), const struct lyd_value *original, struct lyd_value *dup)
+{
+    /* The duplicate is not created here, but at the caller, which creates a duplicate lyds tree
+     * implicitly by inserting duplicate nodes into the data tree.
+     */
+    memset(dup, 0, sizeof *dup);
+    dup->dyn_mem = NULL;
```
Comment (michalvasko) [inline src/plugins_types/lyds_tree.c:89]: Redundant.

File: src/plugins_types/lyds_tree.c [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+static void
+lyplg_type_free_lyds(const struct ly_ctx *UNUSED(ctx), struct lyd_value *value)
+{
+    struct lyd_value_lyds_tree *val = NULL;
+
+    /* The canonical value is not used at all. */
+    assert(!value->_canonical);
+    LYD_VALUE_GET(value, val);
+    /* Release Red-black tree. */
```
Comment (michalvasko) [inline src/plugins_types/lyds_tree.c:77]: Please try to stick to the rule of having an empty line before every comment (unless there is `{` or something similar).

File: src/tree_data.c [hunk 9]
```diff
@@ -989,6 +1088,56 @@ lyd_insert_meta(struct lyd_node *parent, struct lyd_meta *meta, ly_bool clear_df
...
+lyd_get_meta_annotation(const struct lys_module *mod, const char *name, size_t name_len)
+{
+    LY_ARRAY_COUNT_TYPE u;
+
+    if (!mod) {
+        return NULL;
+    }
+
+    LY_ARRAY_FOR(mod->compiled->exts, u) {
+        if (!strncmp(mod->compiled->exts[u].def->plugin->id, "ly2 metadata", 12) &&
```
Comment (michalvasko) [inline src/tree_data.c:1131]: I do not think every extension instance must have a `plugin`, add a check.

File: src/tree_data_sorted.c [hunk 10]
```diff
@@ -0,0 +1,958 @@
...
+    const struct lyd_node *k1, *k2;
+    struct lyd_value *val1, *val2;
+    int cmp;
+
+    /* compare first list key */
+    assert(n1->schema->nodetype & LYS_LIST);
+    assert(n2->schema->nodetype & LYS_LIST);
+    cmp = 0;
+    k1 = ((const struct lyd_node_inner *)n1)->child;
+    k2 = ((const struct lyd_node_inner *)n2)->child;
```
Comment (michalvasko) [inline src/tree_data_sorted.c:226]: Instead use `lyd_child(n1)`.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/tree_data_sorted.c (hunk 10 above)
Comment (lePici) [reply src/tree_data_sorted.c:226]: The `perf` tool complained that `lyd_child()` was using too much CPU.

**Review by michalvasko — APPROVED** [review]

**Pull request merged**